### PR TITLE
Fast test framework + L1–L4 test suites (+ orthogonal perf/QA)

### PR DESCRIPTION
**Pull request by gigara**
## What & why

Stands up a **fast, deterministic test framework** for the extension + webviews, where most of the reported bugs actually live (form/editor rendering, diagram semantics, LS-response transforms) but which had little or no fast-test coverage. The fast suite (**L1–L2**) runs on every PR with no VSCode/LS spawn; heavier tiers (**L3** LS-integration, **L4** E2E) plus the **orthogonal perf/QA** suite are opt-in and distro-gated.

## Test hierarchy

Four **scope levels**, ordered by how much of the system each test actually runs, plus one **orthogonal** suite (perf/QA) — performance is a non-functional dimension that cuts across the pyramid, so it isn't a numbered level. There is ≥1 test at every level.

| Level | What | Tier |
|---|---|---|
| **L1** unit / contract | pure-logic utils (side-panel value/validation, visualizer diagnostics/range/convertConfig, TypeBrowser search) **+ auto-discovered enum contract invariants** over `@wso2/ballerina-core` (integrity + wire-safety) | fast/PR |
| **L2** component | render semantics — 14 side-panel editors + components, EditorFactory selection invariant, bi/sequence diagram semantics, capture-based real-data render+snapshot | fast/PR |
| **L3** LS-integration | headless LS over stdio (flow model, node-template invariants) + capture/drift checks | nightly (needs distro) |
| **L4** E2E | existing Playwright smoke | matrix |
| _orthogonal_ · **perf / QA** | LS latency baseline, visual/a11y | manual |

## Foundation

- **`@wso2/test-config`** — shared Jest preset (jsdom, transforms, asset mocks, single-React pin) + `loadFixture(s)`, and the **headless LS harness** (`ls-harness`, imported as `@wso2/test-config/ls-harness`). Adopt in ~3 lines.
- **Fixture recorder** (`fixtureRecorder.ts`) — env-gated (`BAL_RECORD_FIXTURES`) taps on the LS client + webview RPC; dormant otherwise (zero production impact).
- **Capture script** (`capture-fixtures.js`) — runs the E2E with recording on and dedupes the dump into per-method fixtures.
- **CI `fast-tests` job** — auto-discovers every package with a `jest.config.js`, runs it under xvfb with `--coverage`, and uploads a per-package lcov artifact. Kept fast by the naming rule below.

**The rule that keeps the PR suite fast:** the CI job discovers a package's tests **only** if its config file is named exactly `jest.config.js`. LS/perf/E2E use other config names (`jest.ls-integration.config.js`, `jest.perf.config.js`, `jest.realdata.config.js`, `jest.headless-view.config.js`) so they never run on every PR.

## Patterns (documented in `docs/TEST_GUIDE.md`)

- **The L2 rule:** render the component with the context it needs (props / `renderWithForm` / `renderWithRpc` with a fake client) and feed it **synthetic** data for invariants/edge-cases or a **captured fixture** for fidelity/drift.
- **Real LS data:** capture→file (nightly `*.capture.test.ts` drift-check) + render→snapshot (fast `*.render.test.tsx`). The committed fixture is real rpc data; the fast test never touches the LS.
- **Invariants over per-issue tests** — assert a rule over a corpus; a fix ⇒ a fixture named `issue-<n>.json`.
- **Extract, don't mock-the-world** — a pure function trapped in a heavy module is moved to a narrow-import sibling and re-exported (importers unaffected).

## Notable

- **Fixed a latent production bug:** a circular import in the side-panel editor barrel (`InputMode`) that was order-dependent (worked under webpack, broke under jest).
- **Issue-seeded invariants:** #1491/#105 (array-editor selection), #602/#619 (type-search normalization), #412 (search must not cap results), #766-class (feature-flag node gating), plus `nodeTemplateInvariants` intentionally red until the LS side of #1491 is fixed.
- **Review hardening:** relocated the LS harness into `@wso2/test-config` for a stable public import (no cross-package deep paths); made fixture secret-redaction type-agnostic; added an `error` listener to the LS harness spawn.

## How to run

```bash
# fast suite (per package)
pnpm --filter @wso2/ballerina-side-panel test
# LS-integration / capture (needs Java + a Ballerina distribution; auto-skips otherwise)
cd packages/ballerina-extension && pnpm run test:ls-integration
```

Agent guidance for writing new tests at any layer is in `AGENTS.md` → **Writing tests**.

### CODE REVIEW

File: packages/ballerina-extension/src/test-support/fixtureRecorder.ts [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            return v.map(walk);
+        }
+        const out: any = {};
+        for (const [k, val] of Object.entries(v)) {
+            out[k] = SECRET_KEY_RE.test(k) && typeof val === "string" ? "***REDACTED***" : walk(val);
+        }
+        return out;
+    };
+    return walk(value);
+}
```
[the remark this answers is not in the document]
Comment (gigara) [reply packages/ballerina-extension/src/test-support/fixtureRecorder.ts:72]: Fixed in 4db63d97f. `redactSecrets` now replaces the whole value whenever the key matches `SECRET_KEY_RE`, regardless of type — arrays/objects under `tokens`/`credentials` are no longer recursed into. Added an invariant test (`redacts a secret-looking key wholesale regardless of value type`) to lock it.

File: packages/ballerina-extension/src/test-support/ls-integration/lsHarness.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            env: process.env,
+        });
+        this.proc.stdout.on("data", (d: Buffer) => this.onData(d));
+        this.proc.on("exit", (code) => {
+            for (const [, p] of this.pending) {
+                p.reject(new Error(`LS exited (code ${code}) before responding`));
+            }
+            this.pending.clear();
+        });
+    }
```
[the remark this answers is not in the document]
Comment (gigara) [reply packages/ballerina-extension/src/test-support/ls-integration/lsHarness.ts]: Fixed in 7db3d47a6e. Added an `error` listener on the spawned process that rejects all pending requests and clears state, so a failed `spawn()` surfaces as a normal rejection instead of an unhandled event. (The harness also moved to `@wso2/test-config/ls-harness` — see the capture-test thread — and the fix lives there now.)

File: packages/ballerina-visualizer/src/utils/diagnostics.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+ * under the License.
+ */
+
+// Pure diagnostic-mapping helpers extracted from `bi.tsx` so they can be unit
+// tested without dragging that module's heavy UI dependencies (react-markdown,
+// highlight.js, @wso2/bi-diagram, …). `bi.tsx` re-exports these, so existing
+// importers of `utils/bi` are unaffected. `Diagnostic` is imported as a type only,
+// so this module has no runtime dependency on the @wso2/ballerina-core barrel.
+
+import type { Diagnostic } from "@wso2/ballerina-core";
```
[the remark this answers is not in the document]
Comment (gigara) [reply packages/ballerina-visualizer/src/utils/diagnostics.ts:25]: This one is a false positive, so leaving as-is. The barrel re-exports the LSP type explicitly at `packages/ballerina-core/src/index.ts:127`: `export { Diagnostic } from "vscode-languageserver-types";`. An explicit named re-export overrides the `export *` from `interfaces/bi.ts`, so the `Diagnostic` imported here is the LSP shape (with `range`/`message`) — exactly the contract these helpers use — not the `{ hasDiagnostics, diagnostics }` wrapper. The production `tsc` build compiles this file cleanly, which confirms the resolved type carries `range`/`message`.

File: packages/bi-diagram/src/test/Diagram.flowModel.capture.test.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+// L3-tier: spawns the LS (needs Java + a Ballerina distribution); auto-skips otherwise;
+// runs via jest.realdata.config.js, never the fast PR job.
+
+import * as fs from "fs";
+import * as path from "path";
+import {
+    LsHarness,
+    resolveBalCommand,
+} from "../../../ballerina-extension/src/test-support/ls-integration/lsHarness";
+
```
[the remark this answers is not in the document]
Comment (gigara) [reply packages/bi-diagram/src/test/Diagram.flowModel.capture.test.ts]: Fixed in 7db3d47a6e. `LsHarness`/`resolveBalCommand`/`pathToFileUri` were relocated into the shared `@wso2/test-config` package (as `ls-harness.js` + `ls-harness.d.ts`, matching the existing `fixtures.js` convention). All six consumers — the bi-diagram and record-creator capture tests plus the extension's ls-integration/perf tests — now import from the stable `@wso2/test-config/ls-harness` instead of a deep relative path across package boundaries.

**Review by NipunaRanasinghe — COMMENTED** [review]
@gigara The overall approach and changes LGTM, apart from the below concern about the L0 level.

L0 (enum/contract invariants) effectively looks like a set of unit tests. They run in-process with no I/O, which fits the definition of unit tests.

The only difference is that they assert over an enum table rather than a function. In the usual testing pyramid, "static" refers to checks that don't execute code, such as `tsc` and linting.

So I'd suggest moving these under L1 and labeling them as contract/invariant tests rather than keeping them as a separate level. wdyt?

**Review by NipunaRanasinghe — COMMENTED** [review]
@gigara And on L5: I doubt if perf tests quite fit as a level in the functional pyramid either. The levels there are ordered by scope (how much code/system each test actually runs when it executes), whereas performance is a non-functional dimension that cuts across all of them.

So I'd suggest keeping the numbered levels for the scope pyramid (L1–L4) and listing perf as a separate orthogonal suite rather than L5. wdyt?

**Comment by gigara** [comment]
@NipunaRanasinghe thanks, both points are well taken — agreed and applied in 0d97c8f0ec.

**L0 → L1.** You're right: the enum/contract invariants run in-process and execute code, so calling them "static" was a misnomer (static = `tsc`/lint, which don't run code). They're now **L1 contract tests** — L1 is relabeled *"Unit / contract (pure logic + shared-type invariants)"*, with the enum suite documented as the contract flavour alongside pure-logic units. No separate L0 level anymore.

**Perf → orthogonal.** Also agreed — performance is non-functional and cuts across the scope pyramid, so it's no longer a numbered level. It's now an **orthogonal Perf / QA suite** in the docs (and the `*.perf.ts` self-label updated to match).

Net result is the clean scope pyramid you suggested — **L1 unit/contract → L2 component → L3 LS-integration → L4 E2E**, with Perf/QA orthogonal. L2–L4 numbering is unchanged, so no other references shifted. Updated `docs/TEST_GUIDE.md` + `AGENTS.md` accordingly.

File: packages/test-config/ls-harness.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        this.balCommand = balCommand;
+        this.buf = Buffer.alloc(0);
+        this.nextId = 1;
+        this.pending = new Map();
+        this.defaultTimeoutMs = opts.timeoutMs ?? 60_000;
+    }
+
+    start() {
+        this.proc = spawn(this.balCommand, ["start-language-server"], {
+            stdio: ["pipe", "pipe", "pipe"],
```
Comment (NipunaRanasinghe) [inline packages/test-config/ls-harness.js:72]: The spawned LS has stderr piped but nothing reads it. If the language server writes more than the OS pipe buffer (~64KB) to stderr during a run, it blocks on the write and the test hangs until the timeout — an intermittent CI hang rather than a clear failure. Either set stderr to ignore (`stdio: ["pipe", "pipe", "ignore"]`), or attach a drain listener (`this.proc.stderr.on("data", () => {})`) if you want to keep the logs.

File: packages/ballerina-extension/src/core/extended-language-client.ts [hunk 6]
```diff
@@ -566,6 +567,23 @@ export class ExtendedLangClient extends LanguageClient implements ExtendedLangCl
     }
     init?: (params: InitializeParams) => Promise<InitializeResult>;
 
+    // Records all LS request/response traffic to fixtures when BAL_RECORD_FIXTURES
+    // is set. No-op (and zero overhead) otherwise. See src/test-support/fixtureRecorder.ts.
```
Comment (NipunaRanasinghe) [inline packages/ballerina-extension/src/core/extended-language-client.ts]: This override runs on every LS request in production and reads `process.env.BAL_RECORD_FIXTURES` on each call, so "zero overhead" isn't quite accurate — completions/diagnostics hit this path often. Caching `isRecording()` in a module-level const at load time would make it actually free. Behaviour is fine otherwise.

**Review by NipunaRanasinghe — COMMENTED** [review]
Please find the below minor comments on the test implementation. LGTM other than that. 

- Fixture redaction is key-name based, so a secret embedded in a string value under a non-matching key (e.g. a token inside a URL) won't be masked. Fine given the hook is env-gated and fixtures are curated, but worth stating in the capture workflow that fixtures should be eyeballed before they're committed.
- Test files are now excluded from the build tsconfig and Jest runs with `isolatedModules: true`, so type errors in test code aren't caught by either `rush build` or the test run. Deliberate speed trade, but a `tsc --noEmit` over the tests would close the gap if you want it.

**Comment by gigara** [comment]
@NipunaRanasinghe thanks for the thorough pass — all four addressed:

**1. Redaction is key-name based (secret in a value).** Agreed it stays key-name based; documented the limitation in the capture workflow (`docs/TEST_GUIDE.md`) with an explicit *"review captured fixtures before committing — a secret embedded in a value under a non-matching key is not masked."* callout. (`eb7e480`/`e0a3e37`)

**2. Test code not type-checked (build excludes tests + ts-jest `isolatedModules`).** Added it. Each test package now has a `tsconfig.spec.json` (`extends` its build config, `noEmit`, `include: src/**` for ambient context) plus a shared `@wso2/test-config/jest-matchers.d.ts` ambient (jest-dom matchers + Node globals). `scripts/typecheck-tests.js` runs every package's `tsc --noEmit` and gates on **errors in test files only** — pre-existing production errors (webpack-built packages never `tsc`) and cross-package `src` imports are tallied but don't gate. Wired into the fast-tests CI job. It already caught a real one: `GroupList.test.tsx` passed an `onSearch` prop the component doesn't accept — removed. One pre-existing test (`visualizer/bi.test.ts`, predates this framework) is grandfathered via `exclude` and tracked as a follow-up. (`e0a3e37`)

**3. `extended-language-client` reads `process.env` on every request ("zero overhead" inaccurate).** Fixed — `isRecording()` stays a live read, but `sendRequest` now branches on a module-level const captured once at load, so the not-recording path is genuinely free. Updated the comment to match. (`eb7e480`)

**4. `ls-harness` stderr piped but not drained (pipe-buffer deadlock).** Fixed — attached a drain listener so a chatty LS can't fill the ~64KB buffer and hang the test. (`eb7e480`)

File: scripts/typecheck-tests.js [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    let out = "";
+    try {
+        execFileSync("npx", ["tsc", "--noEmit", "-p", "tsconfig.spec.json"], {
+            cwd,
+            encoding: "utf8",
+            stdio: ["ignore", "pipe", "pipe"],
+        });
+    } catch (e) {
+        out = `${e.stdout || ""}${e.stderr || ""}`;
+    }
```
[the remark this answers is not in the document]
Comment (gigara) [reply scripts/typecheck-tests.js:62]: Good catch — fixed in 32711484c2. The catch now records that tsc exited non-zero, and after parsing, if there are **zero file-scoped diagnostics** (no `path(line,col): error TS…`) it treats the run as failed and reports `typecheck could not run`, incrementing the error count. That covers both a failed spawn (ENOENT, empty output) and a global/config error like `TS18003 "No inputs were found"` — verified both now fail the gate instead of exiting clean. Genuine type errors and clean exits are unaffected.

**Comment by NipunaRanasinghe** [comment]
@coderabbitai review

**Review by NipunaRanasinghe — APPROVED** [review]
LGTM
